Sentiment Analysis for Online Reviews

In [38]:
# libraries to import
import string
import pandas as pd
import nltk as nltk
nltk.download('wordnet')
nltk.download('stopwords')
nltk.download('punkt')
from nltk.corpus import stopwords 
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer  

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\evatr\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\wordnet.zip.
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\evatr\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\evatr\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


a) Downloading, reading and analyzing datasets

In [57]:
# load data in the right> format according to readme files
yelp=pd.read_csv("sentiment_labelled_sentences\yelp_labelled.txt",delimiter="\t", names=["Sentence", "Label"])
imdb=pd.read_csv("sentiment_labelled_sentences\imdb_labelled.txt",delimiter="\t", names=["Sentence", "Label"])
amazon=pd.read_csv("sentiment_labelled_sentences\labelled_amazon.txt",delimiter="\t", names=["Sentence", "Label"])

In [55]:
# check if data is balance in all three dataframes

# yelp
ones_yelp = len(yelp[yelp['Label'] == 1])
zeros_yelp = len(yelp[yelp['Label'] == 0])
print('Number of 1s in Yelp:', onesYelp)
print('Number of 0s in Yelp:', zerosYelp)

#imdb
ones_imdb = len(imdb[imdb['Label'] == 1])
zeros_imdb = len(imdb[imdb['Label'] == 0])
print('Number of 1s in Imdb:', onesImdb)
print('Number of 0s in Imdb:', zerosImdb)

#amazon
ones_amazon = len(amazon[amazon['Label'] == 1])
zeros_amazon = len(amazon[amazon['Label'] == 0])
print('Number of 1s in Amazon:', onesAmazon)
print('Number of 0s in Amazon:', zerosAmazon)

Number of 1s in Yelp: 500
Number of 0s in Yelp: 500
Number of 1s in Imdb: 386
Number of 0s in Imdb: 362
Number of 1s in Amazon: 500
Number of 0s in Amazon: 500


The data in the Yelp and Amazon files is balanced because there are the same number of 1s and 0s as labels. 
The data in the Imdb file can be considered almost balancen because the number of 1s and 0s is almost the same (386 and 362, respectively). The ratio of 1s to 0s is 386/362 = 1.067.

b) Pre-processing datasets

In [58]:
# convert all letters to lower case
yelp = yelp.applymap(lambda s:s.lower() if type(s) == str else s)
imdb = imdb.applymap(lambda s:s.lower() if type(s) == str else s)
amazon = amazon.applymap(lambda s:s.lower() if type(s) == str else s)

# lemmatize, remove punctuation, remove stop words

stop_words = set(stopwords.words('english')) # find stop words in English language
lemmatizer = WordNetLemmatizer() # declare nltk lemmatizer

# iterate through every sentence and replace it by itself lemmatized, without punctuation and without stop words
for i in yelp['Sentence'].index:
    
    # remove punctuation
    sentence_no_punct = ''
    for char in (yelp.at[i, 'Sentence']):
        if char not in string.punctuation:
            sentence_no_punct = sentence_no_punct + char
    (yelp.at[i, 'Sentence']) = sentence_no_punct
    
    # remove stop words
    word_tokens = word_tokenize(yelp.at[i, 'Sentence'])
    (yelp.at[i, 'Sentence']) = ''
    for word in word_tokens: 
        if word not in stop_words: 
            (yelp.at[i, 'Sentence']) += (' ' + lemmatizer.lemmatize(word)) # check why it's not lemmatizing!
            
    # lemmatize words
    #(yelp.at[i, 'Sentence']) = lemmatizer.lemmatize(yelp.at[i, 'Sentence'])
    
print(yelp['Sentence']) # check it worked correctly

0                                        wow loved place
1                                             crust good
2                                    tasty texture nasty
3       stopped late may bank holiday rick steve reco...
4                             selection menu great price
5                            getting angry want damn pho
6                             honeslty didnt taste fresh
7       potato like rubber could tell made ahead time...
8                                              fry great
9                                            great touch
10                                        service prompt
11                                         would go back
12      cashier care ever say still ended wayyy overp...
13       tried cape cod ravoli chickenwith cranberrymmmm
14                      disgusted pretty sure human hair
15                            shocked sign indicate cash
16                                    highly recommended
17                          wai

For this part, we decided to convert all sentences to lower case, so that the same word with some upper case letters and without them would not be detected as different words since we are using the string type which takes into account their differences. We also stripped the sentences of stop words because they do not add any meaning as the same stop words appear in many different sentences. Additionally, we removed the punctuation because it does not add any meaning to the word analysis exercise we will do in this question. We also lemmatized all the words because we are interested in knowing which class of words they belong to in order to understand the meaning of the sentence and not whether they are a noun, adjective, etc.

c) Split training and testing data

In [59]:
# split the three datasets into training and testing data according to the specifications

# split yelp
training_yelp = (yelp.query('Label == 1' )).head(400)
training_yelp.append((yelp.query('Label == 0' )).head(400))
testing_yelp = (yelp.query('Label == 1' )).tail(100)
testing_yelp.append((yelp.query('Label == 0' )).tail(100))

# split imdb
training_imdb = (imdb.query('Label == 1' )).head(400)
training_imdb.append((imdb.query('Label == 0' )).head(400))
testing_imdb = (imdb.query('Label == 1' )).tail(100)
testing_imdb.append((imdb.query('Label == 0' )).tail(100))

# split amazon
training_amazon = (amazon.query('Label == 1' )).head(400)
training_amazon.append((amazon.query('Label == 0' )).head(400))
testing_amazon = (amazon.query('Label == 1' )).tail(100)
testing_amazon.append((amazon.query('Label == 0' )).tail(100))

Sentence  Label
778                              this is a great deal.      1
787                 it is simple to use and i like it.      1
788  it's a great tool for entertainment, communica...      1
791    i own 2 of these cases and would order another.      1
792                                       great phone.      1
793  i bought this battery with a coupon from amazo...      1
795                               perfect for the ps3.      1
796                              five star plus, plus.      1
797  a good quality bargain.. i bought this after i...      1
800                                 good , works fine.      1
801               for the price this was a great deal.      1
802                                  great price, too!      1
804      yes it's shiny on front side - and i love it!      1
807   the reception through this headset is excellent.      1
808                         2 thumbs up to this seller      1
809  i have a verizon lg phone and they work well t...      1
811                #1 it works - #2 it is comfortable.      1
812                 battery charge-life is quite long.      1
813           the jabra eargels fit my ears very well.      1
816                   product is exactly as described.      1
817     gets a signal when other verizon phones won't.      1
819                              i would recommend it.      1
820        plan on ordering from them again and again.      1
821  that being said, for a phone, the camera is ve...      1
822  i have been very satisfied with this cell phon...      1
823  they keep getting better and better (this is m...      1
825             the shipping time was also very quick!      1
827                                        good case!.      1
829                               really good product.      1
850                                it looks very nice.      1
..                                                 ...    ...
959  however, the keypads are so tinny that i somet...      0
965             i cannot make calls at certain places.      0
966  however, my girl was complain that some time t...      0
968  this company charge me a restocking fee and st...      0
969              i'm trying to return it for a refund.      0
970  i plugged it in only to find out not a darn th...      0
972                        earbud piece breaks easily.      0
973                                     lousy product.      0
974  this phone tries very hard to do everything bu...      0
979                                 not enough volume.      0
980                echo problem....very unsatisfactory      0
981  you could only take 2 videos at a time and the...      0
982                            don't waste your money.      0
983  i am going to have to be the first to negative...      0
984  adapter does not provide enough charging current.      0
985  there was so much hype over this phone that i ...      0
986  you also cannot take pictures with it in the c...      0
987                            phone falls out easily.      0
988  it didn't work, people can not hear me when i ...      0
989  the text messaging feature is really tricky to...      0
990  i'm really disappointed all i have now is a ch...      0
991                                painful on the ear.      0
992                   lasted one day and then blew up.      0
993                                      disappointed.      0
994                              kind of flops around.      0
995  the screen does get smudged easily because it ...      0
996  what a piece of junk.. i lose more calls on th...      0
997                       item does not match picture.      0
998  the only thing that disappoint me is the infra...      0
999  you can not answer calls with the unit, never ...      0

[200 rows x 2 columns]

d) Bag of Words model

For this question we cannot use the testing set to create the dictionary of unique words because the model needs to be created with the training set so that we can use the testing set as new data to test our model's ability to generalize. If we create the dictionary with the testing data, we are esentially using all the data as training data and would need to look for another set of new data to test the classifier.

In [70]:
# DID IT ONLY FOR YELP - NEED TO CHECK IF WE SHOULD DO THE 3 OF THEM SEPARATELY OR JUST ONE FOR ALL OF THEM,
# IT IS THE SAME DOUBT IN C)


# create dictionary of unique words in training set
word_dictionary = {}

# iterate through every word or every sentence and store it in dictionary with count 0 (the count will be updated
# later when we iterate through both testing and training set
for i in training_yelp.index:
    word_tokens_training = word_tokenize(training_yelp.at[i, 'Sentence'])
    for word in word_tokens_training:
        if word not in word_dictionary.keys():
            word_dictionary[word] = 0
            
# count the number of occurences of each word in dictionary in training set
for i in training_yelp.index:
    word_tokens_training = word_tokenize(training_yelp.at[i, 'Sentence'])
    for word in word_tokens_training:
        if word in word_dictionary.keys():
            word_dictionary[word] += 1
            
# count the number of occurences of each word in dictionary in testing set  
for i in testing_yelp.index:
    word_tokens_testing = word_tokenize(testing_yelp.at[i, 'Sentence'])
    for word in word_tokens_testing:
        if word in word_dictionary.keys():
            word_dictionary[word] += 1

In [71]:
word_dictionary # just checking here

{'wow': 3,
 'loved': 10,
 'place': 60,
 'stopped': 2,
 'late': 1,
 'may': 3,
 'bank': 1,
 'holiday': 1,
 'rick': 1,
 'steve': 1,
 'recommendation': 3,
 'selection': 10,
 'menu': 13,
 'great': 70,
 'price': 13,
 'fry': 6,
 'touch': 2,
 'service': 46,
 'prompt': 1,
 'tried': 5,
 'cape': 1,
 'cod': 1,
 'ravoli': 1,
 'chickenwith': 1,
 'cranberrymmmm': 1,
 'highly': 3,
 'recommended': 2,
 'food': 60,
 'amazing': 21,
 'also': 18,
 'cute': 2,
 'could': 8,
 'care': 2,
 'le': 1,
 'interior': 1,
 'beautiful': 3,
 'performed': 1,
 'thats': 1,
 'rightthe': 1,
 'red': 1,
 'velvet': 1,
 'cakeohhh': 1,
 'stuff': 2,
 'good': 73,
 'hole': 1,
 'wall': 3,
 'mexican': 2,
 'street': 1,
 'taco': 4,
 'friendly': 23,
 'staff': 15,
 'combo': 1,
 'like': 18,
 'burger': 6,
 'beer': 8,
 '23': 1,
 'decent': 1,
 'deal': 4,
 'found': 3,
 'accident': 1,
 'happier': 1,
 'overall': 4,
 'lot': 5,
 'redeeming': 1,
 'quality': 5,
 'restaurant': 17,
 'inexpensive': 2,
 'ample': 1,
 'portion': 5,
 'first': 13,
 'visit': 4,